packages/graphql-codegen-cli/src/utils/watcher.ts

```diff
@@ -95,6 +96,12 @@ export const createWatcher = (
         // it doesn't matter what has changed, need to run whole process anyway
         await Promise.all(
           events.map(async ({ type: eventName, path }) => {
+            /**
+             * @parcel/watcher has no way to run watcher on specific files (https://github.com/parcel-bundler/watcher/issues/42)
+             * But we can use micromatch to filter out events that we don't care about
+             */
+            if (!mm.contains(path, files)) return;
```

Comment:
@saihaj I believe there's still an issue here when a negated path is used.

My use case looks like this:

```yaml
documents:
  - './apps/frontend/src/app/**/*.graphql'
  - '!./apps/frontend/src/app/exclude-these/**/*.graphql'
```

I'm seeing the watcher run when any file in my source tree is changed, and I'm thinking it's due to `'!./apps/frontend/src/app/exclude-these/**/*.graphql'` causing the `mm.contains` call to return true.

Quick example with mm:
```javascript
const mm = require('micromatch');

// Prints false
console.log(mm.contains('anyfile.js', './some-dir/**/*.graphql'));

// Prints true
console.log(mm.contains('anyfile.js', '!./some-dir/**/*.graphql'));
```

Reply:
I haven't checked the code, but have noticed an excessive amount of runs in watch mode since the update myself.

Reply:
If you can open issue with repro will be easier for me to debug and fix

Reply:
I have downgraded to 3.1.0 and noticed that there is indeed a regression in newer versions causing the generator to run 5-10 times in a row (probably tracking changes in the `.next` folder). For now I had to remove all negated paths and be more specific with included directories which helped to reduce the amount of compilations.

Unfortunantely my project is not public and I don't necesarily have sufficient time to reduce it to a MWE, but I could consider sharing the repository with you @saihaj in case you'd be willing to look into it unless somebody has a decent reproduction.

Reply:
A minor update: I found the `--debug` flag and confirmed that having negations causes them to be picked up causing the excess updates:
![CleanShot 2023-03-11 at 16 46 27](https://user-images.githubusercontent.com/45569997/224494044-d0ea58d8-3a43-43bc-90b7-bcf9150e2450.png)

Reply:
I did a similar investigation by using `--debug` and modifying the source directly and can confirm the same thing that Christian is seeing. I was able to remove the negated pattern and now things work as I want them do. 

Fortunately, our project doesn't need the negated pattern - I only added the negated pattern with v3.2.0 to try to fix the issue that this PR ended up fixing. I am sharing this to confirm the behavior that the other user is also seeing.

Reply:
That's a great find! Let's use this cause this PR notif got hidden in my inbox https://github.com/dotansimha/graphql-code-generator/issues/9177

Reply:
We have the exact same issue in our Next app, now tracking it here #9270 (instead of #9177)

Think I found the issue/root cause, working on a fix now